binaries/coordinator/src/lib.rs

```diff
@@ -269,16 +269,21 @@ async fn start_inner(
                 DataflowEvent::DataflowFinishedOnMachine { machine_id, result } => {
                     match running_dataflows.entry(uuid) {
                         std::collections::hash_map::Entry::Occupied(mut entry) => {
-                            // Archive finished dataflow
-                            archived_dataflows
-                                .entry(uuid)
-                                .or_insert_with(|| ArchivedDataflow::from(entry.get()));
-                            entry.get_mut().machines.remove(&machine_id);
-                            dataflow_results
-                                .entry(uuid)
-                                .or_default()
-                                .insert(machine_id, result);
-                            if entry.get_mut().machines.is_empty() {
+                            let dataflow = entry.get_mut();
+                            dataflow.machines.remove(&machine_id);
+                            println!(
+                                "removed machine id: {machine_id} from dataflow: {:#?}",
+                                dataflow.uuid
+                            );
+                            if dataflow.machines.is_empty() {
+                                // Archive finished dataflow
+                                archived_dataflows
+                                    .entry(uuid)
+                                    .or_insert_with(|| ArchivedDataflow::from(entry.get()));
+                                dataflow_results
+                                    .entry(uuid)
+                                    .or_default()
+                                    .insert(machine_id, result);
```

Comment:
Could you clarify why you moved this statement into the `if` body too? We want to store the results of all machines, not just the last one.

Reply:
Ok, so the thing is that before we might report that a dataflow is finished when the first daemon send a finish message, but actually other daemon has not yet finished. So I guess it would be better to only store one result which might be the aggregation of all result.

Reply:
Edit: Actually, you're right I should only move the archive_dataflow within the if statement.